app/grandchallenge/algorithms/forms.py

```diff
@@ -365,21 +365,77 @@ class Meta:
             "workstation_config": "Viewer Configuration",
         }
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    def __init__(self, *args, user, **kwargs):
+        super().__init__(*args, user=user, **kwargs)
+        self._user = user
 
         self.fields["contact_email"].required = True
         self.fields["display_editors"].required = True
 
-        self.fields["job_requires_gpu_type"].choices = [
-            (choice.value, choice.label)
-            for choice in [GPUTypeChoices.NO_GPU, GPUTypeChoices.T4]
-        ]
+        self.fields["job_requires_gpu_type"].choices = (
+            self.get_selectable_gpu_type_choices()
+        )
         self.fields["job_requires_memory_gb"].validators = [
             MinValueValidator(settings.ALGORITHMS_MIN_MEMORY_GB),
-            MaxValueValidator(settings.ALGORITHMS_MAX_MEMORY_GB),
+            MaxValueValidator(self.get_maximum_settable_memory_gb()),
         ]
 
+    @cached_property
+    def relevant_phases(self):
+        qs = get_objects_for_user(
+            self._user, "evaluation.create_phase_submission"
+        )
+        inputs = self.instance.inputs.all()
+        outputs = self.instance.outputs.all()
+        return qs.annotate(
+            total_algorithm_input_count=Count(
+                "algorithm_inputs", distinct=True
+            ),
+            total_algorithm_output_count=Count(
+                "algorithm_outputs", distinct=True
+            ),
+            relevant_algorithm_input_count=Count(
+                "algorithm_inputs",
+                filter=Q(algorithm_inputs__in=inputs),
+                distinct=True,
+            ),
+            relevant_algorithm_output_count=Count(
+                "algorithm_outputs",
+                filter=Q(algorithm_outputs__in=outputs),
+                distinct=True,
+            ),
+        ).filter(
+            submission_kind=SubmissionKindChoices.ALGORITHM,
+            public=True,
+            total_algorithm_input_count=len(inputs),
+            total_algorithm_output_count=len(outputs),
+            relevant_algorithm_input_count=len(inputs),
+            relevant_algorithm_output_count=len(outputs),
+        )
+
+    def get_selectable_gpu_type_choices(self):
+        choices = [GPUTypeChoices.NO_GPU, GPUTypeChoices.T4]
+        choices.extend(
+            chain.from_iterable(
+                self.relevant_phases.order_by()
```

Comment:
Not using it anymore now, since I replaced it with ArrayAgg, but just for future reference:
I ran into the issue where distinct() was still returning duplicate values. According to the documentation: 
> If you order by fields from a related model, those fields will be added to the selected columns and they may make otherwise duplicate rows appear to be distinct. 

Adding the `order_by()` call without any arguments removes any ordering (including the default ordering). 

This solved the issue of still getting duplicate values. The SQL query changed from this: 
```sql
SELECT DISTINCT "evaluation_phase"."algorithm_selectable_gpu_type_choices",
       "challenges_challenge"."id",
       "evaluation_phase"."submissions_open_at",
       "evaluation_phase"."created"
  FROM "evaluation_phase"
  LEFT OUTER JOIN "evaluation_phase_algorithm_inputs"
    ON ("evaluation_phase"."id" = "evaluation_phase_algorithm_inputs"."phase_id")
  LEFT OUTER JOIN "evaluation_phase_algorithm_outputs"
    ON ("evaluation_phase"."id" = "evaluation_phase_algorithm_outputs"."phase_id")
 INNER JOIN "challenges_challenge"
    ON ("evaluation_phase"."challenge_id" = "challenges_challenge"."id")
 WHERE ("evaluation_phase"."public" AND "evaluation_phase"."submission_kind" = 3)
 GROUP BY "evaluation_phase"."id",
          "challenges_challenge"."id"
HAVING (COUNT(DISTINCT "evaluation_phase_algorithm_inputs"."componentinterface_id") FILTER (WHERE ("evaluation_phase_algorithm_inputs"."componentinterface_id" IN (SELECT U0."id" FROM "components_componentinterface" U0 INNER JOIN "algorithms_algorithm_inputs" U1 ON (U0."id" = U1."componentinterface_id") WHERE U1."algorithm_id" = '48d17a78e3c8433ca2b14f4932803cfe'::uuid))) = 0 AND COUNT(DISTINCT "evaluation_phase_algorithm_outputs"."componentinterface_id") FILTER (WHERE ("evaluation_phase_algorithm_outputs"."componentinterface_id" IN (SELECT U0."id" FROM "components_componentinterface" U0 INNER JOIN "algorithms_algorithm_outputs" U1 ON (U0."id" = U1."componentinterface_id") WHERE U1."algorithm_id" = '48d17a78e3c8433ca2b14f4932803cfe'::uuid))) = 0 AND COUNT(DISTINCT "evaluation_phase_algorithm_inputs"."componentinterface_id") = 0 AND COUNT(DISTINCT "evaluation_phase_algorithm_outputs"."componentinterface_id") = 0)
```

to this:
```sql
SELECT DISTINCT "evaluation_phase"."algorithm_selectable_gpu_type_choices"
  FROM "evaluation_phase"
  LEFT OUTER JOIN "evaluation_phase_algorithm_inputs"
    ON ("evaluation_phase"."id" = "evaluation_phase_algorithm_inputs"."phase_id")
  LEFT OUTER JOIN "evaluation_phase_algorithm_outputs"
    ON ("evaluation_phase"."id" = "evaluation_phase_algorithm_outputs"."phase_id")
 WHERE ("evaluation_phase"."public" AND "evaluation_phase"."submission_kind" = 3)
 GROUP BY "evaluation_phase"."id"
HAVING (COUNT(DISTINCT "evaluation_phase_algorithm_inputs"."componentinterface_id") FILTER (WHERE ("evaluation_phase_algorithm_inputs"."componentinterface_id" IN (SELECT U0."id" FROM "components_componentinterface" U0 INNER JOIN "algorithms_algorithm_inputs" U1 ON (U0."id" = U1."componentinterface_id") WHERE U1."algorithm_id" = 'c9eecb91a8fc49d3b7fa8a701ac3984e'::uuid))) = 0 AND COUNT(DISTINCT "evaluation_phase_algorithm_outputs"."componentinterface_id") FILTER (WHERE ("evaluation_phase_algorithm_outputs"."componentinterface_id" IN (SELECT U0."id" FROM "components_componentinterface" U0 INNER JOIN "algorithms_algorithm_outputs" U1 ON (U0."id" = U1."componentinterface_id") WHERE U1."algorithm_id" = 'c9eecb91a8fc49d3b7fa8a701ac3984e'::uuid))) = 0 AND COUNT(DISTINCT "evaluation_phase_algorithm_inputs"."componentinterface_id") = 0 AND COUNT(DISTINCT "evaluation_phase_algorithm_outputs"."componentinterface_id") = 0)
```

Reply:
Good to know!